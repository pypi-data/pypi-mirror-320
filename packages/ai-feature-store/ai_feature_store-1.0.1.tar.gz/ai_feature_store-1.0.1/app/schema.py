# app/schema.py

from pydantic import BaseModel

class FeatureRequest(BaseModel):
    features: dict  # Expected format: {"Feature Name": Importance Score}
