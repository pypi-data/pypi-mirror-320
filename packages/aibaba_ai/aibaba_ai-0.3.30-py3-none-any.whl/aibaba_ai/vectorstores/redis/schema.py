from typing import TYPE_CHECKING, Any

from langchain._api import create_importer

if TYPE_CHECKING:
    from aiagentsforce_community.vectorstores.redis.schema import (
        FlatVectorField,
        HNSWVectorField,
        NumericFieldSchema,
        RedisDistanceMetric,
        RedisField,
        RedisModel,
        RedisVectorField,
        TagFieldSchema,
        TextFieldSchema,
        read_schema,
    )

# Create a way to dynamically look up deprecated imports.
# Used to consolidate logic for raising deprecation warnings and
# handling optional imports.
DEPRECATED_LOOKUP = {
    "RedisDistanceMetric": "aiagentsforce_community.vectorstores.redis.schema",
    "RedisField": "aiagentsforce_community.vectorstores.redis.schema",
    "TextFieldSchema": "aiagentsforce_community.vectorstores.redis.schema",
    "TagFieldSchema": "aiagentsforce_community.vectorstores.redis.schema",
    "NumericFieldSchema": "aiagentsforce_community.vectorstores.redis.schema",
    "RedisVectorField": "aiagentsforce_community.vectorstores.redis.schema",
    "FlatVectorField": "aiagentsforce_community.vectorstores.redis.schema",
    "HNSWVectorField": "aiagentsforce_community.vectorstores.redis.schema",
    "RedisModel": "aiagentsforce_community.vectorstores.redis.schema",
    "read_schema": "aiagentsforce_community.vectorstores.redis.schema",
}

_import_attribute = create_importer(__package__, deprecated_lookups=DEPRECATED_LOOKUP)


def __getattr__(name: str) -> Any:
    """Look up attributes dynamically."""
    return _import_attribute(name)


__all__ = [
    "RedisDistanceMetric",
    "RedisField",
    "TextFieldSchema",
    "TagFieldSchema",
    "NumericFieldSchema",
    "RedisVectorField",
    "FlatVectorField",
    "HNSWVectorField",
    "RedisModel",
    "read_schema",
]
