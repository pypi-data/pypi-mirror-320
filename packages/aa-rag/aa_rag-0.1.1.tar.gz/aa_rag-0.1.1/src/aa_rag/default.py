# vector db path
from aa_rag.gtypes.enums import IndexType, EmbeddingModel, DBMode, RetrieveType

VECTOR_DB_PATH = "./lancedb"

# index type
INDEX_TYPE = IndexType.CHUNK

# embedding model
EMBEDDING_MODEL = EmbeddingModel.TEXT_EMBEDDING_3_SMALL

# chunk size
INDEX_CHUNK_SIZE = 256

# chunk overlap
INDEX_OVERLAP_SIZE = 100

# db mode
DB_MODE = DBMode.DEINSERT

# retrieve type
RETRIEVE_TYPE = RetrieveType.HYBRID
# retrieve top k
RETRIEVE_TOP_K = 3
#  hybrid retrieve weight
RETRIEVE_HYBRID_DENSE_WEIGHT = 0.5
RETRIEVE_HYBRID_SPARSE_WEIGHT = 0.5
# retrieve return only page content
RETRIEVE_ONLY_PAGE_CONTENT = False
