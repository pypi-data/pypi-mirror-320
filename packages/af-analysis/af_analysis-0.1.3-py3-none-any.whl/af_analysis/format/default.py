#!/usr/bin/env python3
# coding: utf-8

import os
import re
import logging
from tqdm.auto import tqdm
import pandas as pd

logger = logging.getLogger()
logger.setLevel(logging.INFO)

weigths = [
    "alphafold2",
    "alphafold2_ptm",
    "alphafold2_multimer_v1",
    "alphafold2_multimer_v2",
    "alphafold2_multimer_v3",
]


def read_dir(directory):
    """Extract pdb list from a directory."""

    logger.info(f"Reading {directory}")

    log_dict_list = []

    for file in os.listdir(directory):
        if file.endswith(".pdb"):
            token = file.split("_")

            for i, t in enumerate(token):
                if t in ["relaxed", "unrelaxed"]:
                    start_index = i
                    # print(start_index)
                    break

            name_token = "_".join(token[:start_index])
            state = token[start_index]
            rank = int(token[start_index + 2])
            weight = "_".join(token[start_index + 3 : -4])
            assert weight in weigths
            model = int(token[-3])
            seed = int(token[-1][:-4])

            log_dict_list.append(
                {
                    "pdb": os.path.join(directory, file),
                    "query": name_token,
                    "rank": rank,
                    "state": state,
                    "seed": seed,
                    "model": model,
                    "weight": weight,
                }
            )

    log_pd = pd.DataFrame(log_dict_list)
    return log_pd


def add_json(log_pd, directory):
    """Find json files in the directory.

    Parameters
    ----------
    log_pd : pandas.DataFrame
        Dataframe containing the information extracted from the `log.txt` file.
    directory : str
        Path to the directory containing the json files.

    Returns
    -------
    None
        The `log_pd` dataframe is modified in place.
    """

    logger.info(f"Extracting json files location")

    raw_list = os.listdir(directory)
    file_list = []
    for file in raw_list:
        if file.endswith(".json"):
            file_list.append(file)

    json_list = []

    # Get the last recycle:
    if "recycle" not in log_pd.columns:
        last_recycle = log_pd.groupby(["query", "seed", "model", "weight", "state"])
    else:
        last_recycle = (
            log_pd.groupby(["query", "seed", "model", "weight"])["recycle"].transform(
                "max"
            )
            == log_pd["recycle"]
        )

    for i, last in tqdm(enumerate(last_recycle), total=len(last_recycle)):
        row = log_pd.iloc[i]

        reg = rf"{row['query']}_scores_.*_{row['weight']}_model_{row['model']}_seed_{row['seed']:03d}\.json"
        r = re.compile(reg)

        res = list(filter(r.match, file_list))

        if len(res) == 1:
            json_list.append(os.path.join(directory, res[0]))
        elif len(res) == 0:
            logger.warning(f"Not founded : {reg}")
            json_list.append(None)
        else:
            logger.warning(f"Multiple json file for {reg}: {res}")
            json_list.append(res[0])

    log_pd.loc[:, "data_file"] = json_list
