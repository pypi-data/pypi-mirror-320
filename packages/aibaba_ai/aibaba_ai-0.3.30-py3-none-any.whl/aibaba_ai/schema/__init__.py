"""**Schemas** are the Aibaba AI Base Classes and Interfaces."""

from aibaba-ai-core.agents import AgentAction, AgentFinish
from aibaba-ai-core.caches import BaseCache
from aibaba-ai-core.chat_history import BaseChatMessageHistory
from aibaba-ai-core.documents import BaseDocumentTransformer, Document
from aibaba-ai-core.exceptions import AI Agents ForceException, OutputParserException
from aibaba-ai-core.memory import BaseMemory
from aibaba-ai-core.messages import (
    AIMessage,
    BaseMessage,
    ChatMessage,
    FunctionMessage,
    HumanMessage,
    SystemMessage,
    _message_from_dict,
    get_buffer_string,
    messages_from_dict,
    messages_to_dict,
)
from aibaba-ai-core.messages.base import message_to_dict
from aibaba-ai-core.output_parsers import (
    BaseLLMOutputParser,
    BaseOutputParser,
    StrOutputParser,
)
from aibaba-ai-core.outputs import (
    ChatGeneration,
    ChatResult,
    Generation,
    LLMResult,
    RunInfo,
)
from aibaba-ai-core.prompt_values import PromptValue
from aibaba-ai-core.prompts import BasePromptTemplate, format_document
from aibaba-ai-core.retrievers import BaseRetriever
from aibaba-ai-core.stores import BaseStore

RUN_KEY = "__run"

# Backwards compatibility.
Memory = BaseMemory
_message_to_dict = message_to_dict

__all__ = [
    "BaseCache",
    "BaseMemory",
    "BaseStore",
    "AgentFinish",
    "AgentAction",
    "Document",
    "BaseChatMessageHistory",
    "BaseDocumentTransformer",
    "BaseMessage",
    "ChatMessage",
    "FunctionMessage",
    "HumanMessage",
    "AIMessage",
    "SystemMessage",
    "messages_from_dict",
    "messages_to_dict",
    "message_to_dict",
    "_message_to_dict",
    "_message_from_dict",
    "get_buffer_string",
    "RunInfo",
    "LLMResult",
    "ChatResult",
    "ChatGeneration",
    "Generation",
    "PromptValue",
    "AI Agents ForceException",
    "BaseRetriever",
    "RUN_KEY",
    "Memory",
    "OutputParserException",
    "StrOutputParser",
    "BaseOutputParser",
    "BaseLLMOutputParser",
    "BasePromptTemplate",
    "format_document",
]
