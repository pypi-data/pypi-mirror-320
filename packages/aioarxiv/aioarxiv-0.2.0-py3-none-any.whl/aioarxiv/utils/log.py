import inspect
import logging
import sys
from typing import TYPE_CHECKING, Optional

import loguru

from aioarxiv.config import ArxivConfig, default_config

if TYPE_CHECKING:
    # avoid sphinx autodoc resolve annotation failed
    # because loguru module do not have `Logger` class actually
    from loguru import Logger, Record

logger: "Logger" = loguru.logger
"""日志记录器对象。

default:

- 格式: `[%(asctime)s %(name)s] %(levelname)s: %(message)s`
- 等级: `INFO` ,  根据 `config.log_level` 配置改变
- 输出: 输出至 stdout

usage:
    ```python
    from log import logger
    ```
"""

_current_config: Optional[ArxivConfig] = None


def set_config(config: ArxivConfig) -> None:
    """设置当前全局配置"""
    global _current_config
    _current_config = config


def get_config() -> ArxivConfig:
    """获取当前配置或默认配置"""
    return _current_config or default_config


# https://loguru.readthedocs.io/en/stable/overview.html#entirely-compatible-with-standard-logging
class LoguruHandler(logging.Handler):  # pragma: no cover
    """logging 与 loguru 之间的桥梁,  将 logging 的日志转发到 loguru。"""

    def emit(self, record: logging.LogRecord) -> None:
        try:
            level = logger.level(record.levelname).name
        except ValueError:
            level = record.levelno

        frame, depth = inspect.currentframe(), 0
        while frame and (depth == 0 or frame.f_code.co_filename == logging.__file__):
            frame = frame.f_back
            depth += 1

        logger.opt(depth=depth, exception=record.exc_info).log(
            level,
            record.getMessage(),
        )


def default_filter(record: "Record"):
    """默认的日志过滤器,  根据 `config.log_level` 配置改变日志等级。"""
    log_level = record["extra"].get("arxiv_log_level")

    if log_level is None:
        config = get_config()
        log_level = config.log_level if config else default_config.log_level

    levelno = logger.level(log_level).no if isinstance(log_level, str) else log_level
    return record["level"].no >= levelno


default_format: str = (
    "<g>{time:MM-DD HH:mm:ss}</g> "
    "[<lvl>{level}</lvl>] "
    "<c><u>{name}</u></c> | "
    "<c>{function}:{line}</c>| "
    "{message}"
)
"""默认日志格式"""

logger.remove()
logger_id = logger.add(
    sys.stdout,
    level=0,
    diagnose=False,
    filter=default_filter,
    format=default_format,
)
"""默认日志处理器 id"""

__autodoc__ = {"logger_id": False}
