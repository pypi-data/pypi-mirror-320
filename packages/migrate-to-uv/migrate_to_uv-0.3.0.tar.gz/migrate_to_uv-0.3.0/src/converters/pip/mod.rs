mod dependencies;

use crate::converters::pyproject_updater::PyprojectUpdater;
use crate::converters::DependencyGroupsStrategy;
use crate::converters::{lock_dependencies, Converter};
use crate::schema::pep_621::Project;
use crate::schema::pyproject::DependencyGroupSpecification;
use crate::schema::uv::Uv;
use crate::toml::PyprojectPrettyFormatter;
use indexmap::IndexMap;
use log::{info, warn};
use owo_colors::OwoColorize;
#[cfg(test)]
use std::any::Any;
use std::default::Default;
use std::fs;
use std::fs::{remove_file, File};
use std::io::Write;
use std::path::{Path, PathBuf};
use toml_edit::visit_mut::VisitMut;
use toml_edit::DocumentMut;

#[derive(Debug, PartialEq, Eq)]
pub struct Pip {
    pub project_path: PathBuf,
    pub requirements_files: Vec<String>,
    pub dev_requirements_files: Vec<String>,
    pub is_pip_tools: bool,
}

impl Converter for Pip {
    fn convert_to_uv(
        &self,
        dry_run: bool,
        skip_lock: bool,
        keep_old_metadata: bool,
        _dependency_groups_strategy: DependencyGroupsStrategy,
    ) {
        let pyproject_path = self.project_path.join("pyproject.toml");
        let updated_pyproject_string = perform_migration(
            &self.project_path,
            self.requirements_files.clone(),
            self.dev_requirements_files.clone(),
            &pyproject_path,
        );

        if dry_run {
            info!(
                "{}\n{}",
                "Migrated pyproject.toml:".bold(),
                updated_pyproject_string
            );
        } else {
            let mut pyproject_file = File::create(&pyproject_path).unwrap();

            pyproject_file
                .write_all(updated_pyproject_string.as_bytes())
                .unwrap();

            if !keep_old_metadata {
                delete_requirements_files(&self.project_path, &self.requirements_files).unwrap();
                delete_requirements_files(&self.project_path, &self.dev_requirements_files)
                    .unwrap();

                if self.is_pip_tools {
                    delete_requirements_files(
                        &self.project_path,
                        &self
                            .requirements_files
                            .iter()
                            .map(|file| file.replace(".in", ".txt"))
                            .collect(),
                    )
                    .unwrap();
                    delete_requirements_files(
                        &self.project_path,
                        &self
                            .dev_requirements_files
                            .iter()
                            .map(|file| file.replace(".in", ".txt"))
                            .collect(),
                    )
                    .unwrap();
                }
            }

            if !dry_run && !skip_lock && lock_dependencies(self.project_path.as_ref()).is_err() {
                warn!(
                    "Project migrated from {} to uv, but an error occurred when locking dependencies.",
                    if self.is_pip_tools {
                        "pip-tools"
                    } else {
                        "pip"
                    }
                );
                return;
            }

            info!(
                "{}",
                format!(
                    "Successfully migrated project from {} to uv!\n",
                    if self.is_pip_tools {
                        "pip-tools"
                    } else {
                        "pip"
                    }
                )
                .bold()
                .green()
            );
        }
    }

    #[cfg(test)]
    fn as_any(&self) -> &dyn Any {
        self
    }
}

fn perform_migration(
    project_path: &Path,
    requirements_files: Vec<String>,
    dev_requirements_files: Vec<String>,
    pyproject_path: &Path,
) -> String {
    let dev_dependencies = dependencies::get(project_path, dev_requirements_files);

    let dependency_groups = dev_dependencies.map_or_else(
        || None,
        |dependencies| {
            let mut groups = IndexMap::new();

            groups.insert(
                "dev".to_string(),
                dependencies
                    .iter()
                    .map(|dep| DependencyGroupSpecification::String(dep.to_string()))
                    .collect(),
            );

            Some(groups)
        },
    );

    let project = Project {
        // "name" is required by uv.
        name: Some(String::new()),
        // "version" is required by uv.
        version: Some("0.0.1".to_string()),
        dependencies: dependencies::get(project_path, requirements_files),
        ..Default::default()
    };

    let uv = Uv {
        package: Some(false),
        ..Default::default()
    };

    let pyproject_toml_content = fs::read_to_string(pyproject_path).unwrap_or_default();
    let mut updated_pyproject = pyproject_toml_content.parse::<DocumentMut>().unwrap();
    let mut pyproject_updater = PyprojectUpdater {
        pyproject: &mut updated_pyproject,
    };

    pyproject_updater.insert_pep_621(&project);
    pyproject_updater.insert_dependency_groups(dependency_groups.as_ref());
    pyproject_updater.insert_uv(&uv);

    let mut visitor = PyprojectPrettyFormatter {
        parent_keys: Vec::new(),
    };
    visitor.visit_document_mut(&mut updated_pyproject);

    updated_pyproject.to_string()
}

fn delete_requirements_files(
    project_path: &Path,
    requirements_files: &Vec<String>,
) -> std::io::Result<()> {
    for requirements_file in requirements_files {
        let requirements_path = project_path.join(requirements_file);

        if requirements_path.exists() {
            remove_file(requirements_path.clone())?;
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_perform_pip_tools_migration() {
        insta::assert_toml_snapshot!(perform_migration(
            Path::new("tests/fixtures/pip_tools/full"),
            vec!["requirements.in".to_string()],
            vec!["requirements-dev.in".to_string()],
            Path::new("tests/fixtures/pip_tools/full/pyproject.toml"),
        ));
    }

    #[test]
    fn test_perform_pip_tools_all_files_migration() {
        insta::assert_toml_snapshot!(perform_migration(
            Path::new("tests/fixtures/pip_tools/full"),
            vec!["requirements.in".to_string()],
            vec![
                "requirements-dev.in".to_string(),
                "requirements-typing.in".to_string()
            ],
            Path::new("tests/fixtures/pip_tools/full/pyproject.toml"),
        ));
    }

    #[test]
    fn test_perform_pip_migration() {
        insta::assert_toml_snapshot!(perform_migration(
            Path::new("tests/fixtures/pip/full"),
            vec!["requirements.txt".to_string()],
            vec!["requirements-dev.txt".to_string()],
            Path::new("tests/fixtures/pip/full/pyproject.toml"),
        ));
    }

    #[test]
    fn test_perform_pip_all_files_migration() {
        insta::assert_toml_snapshot!(perform_migration(
            Path::new("tests/fixtures/pip/full"),
            vec!["requirements.txt".to_string()],
            vec![
                "requirements-dev.txt".to_string(),
                "requirements-typing.txt".to_string()
            ],
            Path::new("tests/fixtures/pip/full/pyproject.toml"),
        ));
    }
}
