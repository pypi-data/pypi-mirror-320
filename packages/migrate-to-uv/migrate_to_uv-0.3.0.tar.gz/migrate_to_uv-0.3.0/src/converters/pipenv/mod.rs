mod dependencies;
mod project;
mod sources;

use crate::converters::pyproject_updater::PyprojectUpdater;
use crate::converters::DependencyGroupsStrategy;
use crate::converters::{lock_dependencies, Converter};
use crate::schema::pep_621::Project;
use crate::schema::pipenv::Pipfile;
use crate::schema::uv::{SourceContainer, Uv};
use crate::toml::PyprojectPrettyFormatter;
use indexmap::IndexMap;
use log::{info, warn};
use owo_colors::OwoColorize;
#[cfg(test)]
use std::any::Any;
use std::default::Default;
use std::fs;
use std::fs::{remove_file, File};
use std::io::Write;
use std::path::{Path, PathBuf};
use toml_edit::visit_mut::VisitMut;
use toml_edit::DocumentMut;

#[derive(Debug, PartialEq, Eq)]
pub struct Pipenv {
    pub project_path: PathBuf,
}

impl Converter for Pipenv {
    fn convert_to_uv(
        &self,
        dry_run: bool,
        skip_lock: bool,
        keep_old_metadata: bool,
        dependency_groups_strategy: DependencyGroupsStrategy,
    ) {
        let pipfile_path = self.project_path.join("Pipfile");
        let pyproject_path = self.project_path.join("pyproject.toml");
        let updated_pyproject_string =
            perform_migration(&pipfile_path, &pyproject_path, dependency_groups_strategy);

        if dry_run {
            info!(
                "{}\n{}",
                "Migrated pyproject.toml:".bold(),
                updated_pyproject_string
            );
        } else {
            let mut pyproject_file = File::create(&pyproject_path).unwrap();

            pyproject_file
                .write_all(updated_pyproject_string.as_bytes())
                .unwrap();

            if !keep_old_metadata {
                delete_pipenv_references(&self.project_path).unwrap();
            }

            if !dry_run && !skip_lock && lock_dependencies(self.project_path.as_ref()).is_err() {
                warn!("Project migrated from Pipenv to uv, but an error occurred when locking dependencies.");
                return;
            }

            info!(
                "{}",
                "Successfully migrated project from Pipenv to uv!\n"
                    .bold()
                    .green()
            );
        }
    }

    #[cfg(test)]
    fn as_any(&self) -> &dyn Any {
        self
    }
}

fn perform_migration(
    pipfile_path: &Path,
    pyproject_path: &Path,
    dependency_groups_strategy: DependencyGroupsStrategy,
) -> String {
    let pipfile_content = fs::read_to_string(pipfile_path).unwrap();
    let pipfile: Pipfile = toml::from_str(pipfile_content.as_str()).unwrap();

    let mut uv_source_index: IndexMap<String, SourceContainer> = IndexMap::new();
    let (dependency_groups, uv_default_groups) =
        dependencies::get_dependency_groups_and_default_groups(
            &pipfile,
            &mut uv_source_index,
            dependency_groups_strategy,
        );

    let project = Project {
        // "name" is required by uv.
        name: Some(String::new()),
        // "version" is required by uv.
        version: Some("0.0.1".to_string()),
        requires_python: project::get_requires_python(pipfile.requires),
        dependencies: dependencies::get(pipfile.packages.as_ref(), &mut uv_source_index),
        ..Default::default()
    };

    let uv = Uv {
        package: Some(false),
        index: sources::get_indexes(pipfile.source),
        sources: if uv_source_index.is_empty() {
            None
        } else {
            Some(uv_source_index)
        },
        default_groups: uv_default_groups,
    };

    let pyproject_toml_content = fs::read_to_string(pyproject_path).unwrap_or_default();
    let mut updated_pyproject = pyproject_toml_content.parse::<DocumentMut>().unwrap();
    let mut pyproject_updater = PyprojectUpdater {
        pyproject: &mut updated_pyproject,
    };

    pyproject_updater.insert_pep_621(&project);
    pyproject_updater.insert_dependency_groups(dependency_groups.as_ref());
    pyproject_updater.insert_uv(&uv);

    let mut visitor = PyprojectPrettyFormatter {
        parent_keys: Vec::new(),
    };
    visitor.visit_document_mut(&mut updated_pyproject);

    updated_pyproject.to_string()
}

fn delete_pipenv_references(project_path: &Path) -> std::io::Result<()> {
    let pipfile_path = project_path.join("Pipfile");

    if pipfile_path.exists() {
        remove_file(pipfile_path)?;
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_perform_migration() {
        insta::assert_toml_snapshot!(perform_migration(
            Path::new("tests/fixtures/pipenv/full/Pipfile"),
            Path::new("tests/fixtures/pipenv/full/pyproject.toml"),
            DependencyGroupsStrategy::SetDefaultGroups,
        ));
    }

    #[test]
    fn test_perform_migration_dep_group_include_in_dev() {
        insta::assert_toml_snapshot!(perform_migration(
            Path::new("tests/fixtures/pipenv/full/Pipfile"),
            Path::new("tests/fixtures/pipenv/full/pyproject.toml"),
            DependencyGroupsStrategy::IncludeInDev,
        ));
    }

    #[test]
    fn test_perform_migration_dep_group_keep_existing() {
        insta::assert_toml_snapshot!(perform_migration(
            Path::new("tests/fixtures/pipenv/full/Pipfile"),
            Path::new("tests/fixtures/pipenv/full/pyproject.toml"),
            DependencyGroupsStrategy::KeepExisting,
        ));
    }

    #[test]
    fn test_perform_migration_dep_group_merge_in_dev() {
        insta::assert_toml_snapshot!(perform_migration(
            Path::new("tests/fixtures/pipenv/full/Pipfile"),
            Path::new("tests/fixtures/pipenv/full/pyproject.toml"),
            DependencyGroupsStrategy::MergeIntoDev,
        ));
    }

    #[test]
    fn test_perform_migration_python_full_version() {
        insta::assert_toml_snapshot!(perform_migration(
            Path::new("tests/fixtures/pipenv/python_full_version/Pipfile"),
            Path::new("tests/fixtures/pipenv/python_full_version/pyproject.toml"),
            DependencyGroupsStrategy::SetDefaultGroups,
        ));
    }

    #[test]
    fn test_perform_migration_empty_requires() {
        insta::assert_toml_snapshot!(perform_migration(
            Path::new("tests/fixtures/pipenv/empty_requires/Pipfile"),
            Path::new("tests/fixtures/pipenv/empty_requires/pyproject.toml"),
            DependencyGroupsStrategy::SetDefaultGroups,
        ));
    }

    #[test]
    fn test_perform_migration_minimal_pipfile() {
        insta::assert_toml_snapshot!(perform_migration(
            Path::new("tests/fixtures/pipenv/minimal/Pipfile"),
            Path::new("tests/fixtures/pipenv/minimal/pyproject.toml"),
            DependencyGroupsStrategy::SetDefaultGroups,
        ));
    }
}
