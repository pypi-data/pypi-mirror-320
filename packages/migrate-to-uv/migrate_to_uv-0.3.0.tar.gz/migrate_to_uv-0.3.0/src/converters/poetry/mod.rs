mod build_backend;
mod dependencies;
mod project;
mod sources;
pub mod version;

use crate::converters::poetry::build_backend::get_hatch;
use crate::converters::pyproject_updater::PyprojectUpdater;
use crate::converters::DependencyGroupsStrategy;
use crate::converters::{lock_dependencies, Converter};
use crate::schema::pep_621::Project;
use crate::schema::pyproject::PyProject;
use crate::schema::uv::{SourceContainer, Uv};
use crate::toml::PyprojectPrettyFormatter;
use indexmap::IndexMap;
use log::{info, warn};
use owo_colors::OwoColorize;
#[cfg(test)]
use std::any::Any;
use std::fs;
use std::fs::{remove_file, File};
use std::io::Write;
use std::path::{Path, PathBuf};
use toml_edit::visit_mut::VisitMut;
use toml_edit::DocumentMut;

#[derive(Debug, PartialEq, Eq)]
pub struct Poetry {
    pub project_path: PathBuf,
}

impl Converter for Poetry {
    fn convert_to_uv(
        &self,
        dry_run: bool,
        skip_lock: bool,
        keep_old_metadata: bool,
        dependency_groups_strategy: DependencyGroupsStrategy,
    ) {
        let pyproject_path = self.project_path.join("pyproject.toml");
        let updated_pyproject_string = perform_migration(
            &pyproject_path,
            keep_old_metadata,
            dependency_groups_strategy,
        );

        if dry_run {
            info!(
                "{}\n{}",
                "Migrated pyproject.toml:".bold(),
                updated_pyproject_string
            );
        } else {
            let mut pyproject_file = File::create(&pyproject_path).unwrap();

            pyproject_file
                .write_all(updated_pyproject_string.as_bytes())
                .unwrap();

            if !keep_old_metadata {
                delete_poetry_references(&self.project_path).unwrap();
            }

            if !dry_run && !skip_lock && lock_dependencies(self.project_path.as_ref()).is_err() {
                warn!("Project migrated from Poetry to uv, but an error occurred when locking dependencies.");
                return;
            }

            info!(
                "{}",
                "Successfully migrated project from Poetry to uv!\n"
                    .bold()
                    .green()
            );
        }
    }

    #[cfg(test)]
    fn as_any(&self) -> &dyn Any {
        self
    }
}

fn perform_migration(
    pyproject_path: &Path,
    keep_old_metadata: bool,
    dependency_groups_strategy: DependencyGroupsStrategy,
) -> String {
    let pyproject_toml_content = fs::read_to_string(pyproject_path).unwrap_or_default();
    let pyproject: PyProject = toml::from_str(pyproject_toml_content.as_str()).unwrap();

    let poetry = pyproject.tool.unwrap().poetry.unwrap();

    let mut uv_source_index: IndexMap<String, SourceContainer> = IndexMap::new();
    let (dependency_groups, uv_default_groups) =
        dependencies::get_dependency_groups_and_default_groups(
            &poetry,
            &mut uv_source_index,
            dependency_groups_strategy,
        );
    let mut poetry_dependencies = poetry.dependencies;

    let python_specification = poetry_dependencies
        .as_mut()
        .and_then(|dependencies| dependencies.shift_remove("python"));

    let optional_dependencies = dependencies::get_optional(&mut poetry_dependencies, poetry.extras);

    let mut poetry_plugins = poetry.plugins;
    let scripts_from_plugins = poetry_plugins
        .as_mut()
        .and_then(|plugins| plugins.shift_remove("console_scripts"));
    let gui_scripts = poetry_plugins
        .as_mut()
        .and_then(|plugins| plugins.shift_remove("gui_scripts"));

    let project = Project {
        // "name" is required by uv.
        name: Some(poetry.name.unwrap_or_default()),
        // "version" is required by uv.
        version: Some(poetry.version.unwrap_or_else(|| "0.0.1".to_string())),
        description: poetry.description,
        authors: project::get_authors(poetry.authors),
        requires_python: python_specification.map(|p| p.to_pep_508()),
        readme: project::get_readme(poetry.readme),
        license: poetry.license,
        maintainers: project::get_authors(poetry.maintainers),
        keywords: poetry.keywords,
        classifiers: poetry.classifiers,
        dependencies: dependencies::get(poetry_dependencies.as_ref(), &mut uv_source_index),
        optional_dependencies,
        urls: project::get_urls(
            poetry.urls,
            poetry.homepage,
            poetry.repository,
            poetry.documentation,
        ),
        scripts: project::get_scripts(poetry.scripts, scripts_from_plugins),
        gui_scripts,
        entry_points: poetry_plugins,
    };

    let uv = Uv {
        package: poetry.package_mode,
        index: sources::get_indexes(poetry.source),
        sources: if uv_source_index.is_empty() {
            None
        } else {
            Some(uv_source_index)
        },
        default_groups: uv_default_groups,
    };

    let hatch = get_hatch(
        poetry.packages.as_ref(),
        poetry.include.as_ref(),
        poetry.exclude.as_ref(),
    );

    let mut updated_pyproject = pyproject_toml_content.parse::<DocumentMut>().unwrap();
    let mut pyproject_updater = PyprojectUpdater {
        pyproject: &mut updated_pyproject,
    };

    pyproject_updater
        .insert_build_system(build_backend::get_new_build_system(pyproject.build_system).as_ref());
    pyproject_updater.insert_pep_621(&project);
    pyproject_updater.insert_dependency_groups(dependency_groups.as_ref());
    pyproject_updater.insert_uv(&uv);
    pyproject_updater.insert_hatch(hatch.as_ref());

    if !keep_old_metadata {
        remove_pyproject_poetry_section(&mut updated_pyproject);
    }

    let mut visitor = PyprojectPrettyFormatter {
        parent_keys: Vec::new(),
    };
    visitor.visit_document_mut(&mut updated_pyproject);

    updated_pyproject.to_string()
}

fn remove_pyproject_poetry_section(pyproject: &mut DocumentMut) {
    pyproject
        .get_mut("tool")
        .unwrap()
        .as_table_mut()
        .unwrap()
        .remove("poetry");
}

fn delete_poetry_references(project_path: &Path) -> std::io::Result<()> {
    let poetry_lock_path = project_path.join("poetry.lock");

    if poetry_lock_path.exists() {
        remove_file(poetry_lock_path)?;
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_perform_migration() {
        insta::assert_toml_snapshot!(perform_migration(
            Path::new("tests/fixtures/poetry/full/pyproject.toml"),
            false,
            DependencyGroupsStrategy::SetDefaultGroups,
        ));
    }

    #[test]
    fn test_perform_migration_keep_old_metadata() {
        insta::assert_toml_snapshot!(perform_migration(
            Path::new("tests/fixtures/poetry/full/pyproject.toml"),
            true,
            DependencyGroupsStrategy::SetDefaultGroups,
        ));
    }

    #[test]
    fn test_perform_migration_dep_group_include_in_dev() {
        insta::assert_toml_snapshot!(perform_migration(
            Path::new("tests/fixtures/poetry/full/pyproject.toml"),
            false,
            DependencyGroupsStrategy::IncludeInDev,
        ));
    }

    #[test]
    fn test_perform_migration_dep_group_keep_existing() {
        insta::assert_toml_snapshot!(perform_migration(
            Path::new("tests/fixtures/poetry/full/pyproject.toml"),
            false,
            DependencyGroupsStrategy::KeepExisting,
        ));
    }

    #[test]
    fn test_perform_migration_dep_group_merge_in_dev() {
        insta::assert_toml_snapshot!(perform_migration(
            Path::new("tests/fixtures/poetry/full/pyproject.toml"),
            false,
            DependencyGroupsStrategy::MergeIntoDev,
        ));
    }

    #[test]
    fn test_perform_migration_multiple_readmes() {
        insta::assert_toml_snapshot!(perform_migration(
            Path::new("tests/fixtures/poetry/multiple_readmes/pyproject.toml"),
            false,
            DependencyGroupsStrategy::SetDefaultGroups,
        ));
    }

    #[test]
    fn test_perform_migration_minimal_pyproject() {
        insta::assert_toml_snapshot!(perform_migration(
            Path::new("tests/fixtures/poetry/minimal/pyproject.toml"),
            false,
            DependencyGroupsStrategy::SetDefaultGroups,
        ));
    }
}
