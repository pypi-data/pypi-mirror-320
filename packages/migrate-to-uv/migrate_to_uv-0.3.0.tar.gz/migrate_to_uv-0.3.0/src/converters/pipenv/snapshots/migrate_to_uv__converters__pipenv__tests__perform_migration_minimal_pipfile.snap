---
source: src/converters/pipenv/mod.rs
expression: "perform_migration(Path::new(\"tests/fixtures/pipenv/minimal/Pipfile\"),\nPath::new(\"tests/fixtures/pipenv/minimal/pyproject.toml\"),\nDependencyGroupsStrategy::SetDefaultGroups,)"
snapshot_kind: text
---
'''
[project]
name = ""
version = "0.0.1"

[tool.uv]
package = false
'''
